### DISCUSSION TITLE
ConnectionPool extra parameters

### DISCUSSION THREAD

**Discussion by thrasibule** [psycopg/psycopg#237] [Q&A] [answered]
Just something I stumbled upon while using a ConnectionPool. I'm trying to pass a row_factory to a ConnectionPool, something like that:

```python
from psycopg .rows import namedtuple_row
from psycopg_pool import ConnectionPool
pool = ConnectionPool("postgresql://myserver/mydb", row_factory=namedtuple_row)
```
from the docs I expect row_factory to be bundled up in kwargs, which it is, but then I get this error:

```
---> 54 super().__init__(conninfo, **kwargs)
     56 self._sched_runner = threading.Thread(
     57     target=self._sched.run, name=f"{self.name}-scheduler", daemon=True
     58 )
     59 for i in range(self.num_workers):

TypeError: BasePool.__init__() got an unexpected keyword argument 'row_factory'
```
cause BasePool doesn't take **kwargs as an argument but a literal kwargs, so I think the right call is:
```python
pool = ConnectionPool("postgresql://myserver/mydb", kwargs={"row_factory":namedtuple_row})
```
I feel it would be more natural to have BasePool take the keyword args directly from the derived class. It makes the feature hard to discover otherwise, and would need to be mentioned in the ConnectionPool docs at least.

**Comment by dvarrazzo** [comment, marked as answer]
> and would need to be mentioned in the ConnectionPool docs at least.

It is: [in ConnectionPool init docs](https://www.psycopg.org/psycopg3/docs/api/pool.html#psycopg_pool.ConnectionPool):

> *kwargs* (`dict`) – Extra arguments to pass to `connect()`. Note that this is *one dict argument* of the pool constructor, which is expanded as `connect()` keyword parameters.

> I feel it would be more natural to have BasePool take the keyword args directly from the derived class

Thank you for the suggestion, I appreciate it, but as much as it feels more natural it is not a good idea: the Connection object takes an open-ended set of keyword parameters (because they are passed through to the libpq connection function) and the pool takes a large number of parameters itself: the risk of a clash is too high.

**Reply by thrasibule** [reply]
Thank you I'm not sure how I missed that! I guess I just looked at the function signature. Keeping it in a dict makes a lot of sense for the reason you mentioned.

**Reply by dvarrazzo** [reply]
There was another awkward factor in the design of that signature. Usually you pass **kwargs down the base class to let it manage the arguments the superclass doesn't care about. However, in the pool, the connection is taken by the superclass (because sync and async pools create two different types of connection). So we are in the situation that part of the **kwargs should be passed to the subclass, and some retained.

The set of the retained argument is open-ended as said, so they would be hard to split. The solution would have been to re-define all the arguments of the pool and pass them down explicitly, which might have been even useful from the PoV of documentation, but it's very, uncomfortably, non-DRY.
